app/controllers/automated_tests_controller.rb

```diff
@@ -35,14 +35,34 @@ def student_interface
     @student = current_role
     @grouping = @student.accepted_grouping_for(@assignment.id)
 
-    unless @grouping.nil?
+    if @grouping.present?
       @grouping.refresh_test_tokens
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
+
+      if @assignment.enable_student_tests
+        last_student_run = @grouping.test_runs.where(role: @grouping.accepted_students).first
+        if last_student_run.nil?
+          @next_token_generation_time = @assignment.token_start_date + @assignment.token_period.hours
+        else
+          hours_from_start = (Time.current - @assignment.token_start_date) / 3600
+          if @assignment.non_regenerating_tokens
+            last_period_begin = @assignment.token_start_date
+          else
+            periods_from_start = (hours_from_start / @assignment.token_period).floor
```

Comment:
This calculation is good, but doesn't correctly handle the case when the current time is *before* the assignment token start date. In that case, this value will be negative, but you should set it to `0` instead.